elixir/apps/domain/lib/domain/billing/event_handler.ex

```diff
@@ -1,219 +1,317 @@
-  def handle_event(%{
-        "object" => "event",
-        "data" => %{
-          "object" =>
-            %{
-              "id" => customer_id,
-              "name" => customer_name,
-              "email" => customer_email,
-              "metadata" => customer_metadata
-            } = customer
-        },
-        "type" => "customer.updated"
-      }) do
-    attrs =
-      %{
-        name: customer_metadata["account_name"] || customer_name,
-        legal_name: customer_name,
-        metadata: %{stripe: %{billing_email: customer_email}}
-      }
-      |> put_if_not_nil(:slug, customer_metadata["account_slug"])
+  defp process_event_with_lock(event) do
+    customer_id = extract_customer_id(event)
+    hashed_id = :erlang.phash2(customer_id)
 
-    case update_account_by_stripe_customer_id(customer_id, attrs) do
-      {:ok, _account} ->
-        :ok
+    Repo.transact(fn ->
+      Repo.query!("SELECT pg_advisory_xact_lock($1)", [hashed_id])
+      process_event(event, customer_id)
+    end)
+  end
 
-      {:error, :customer_not_provisioned} ->
-        _ = create_account_from_stripe_customer(customer)
-        :ok
-
-      {:error, :not_found} ->
-        :ok
+  defp process_event(event, customer_id) do
+    with :ok <- check_event_processing_eligibility(event, customer_id),
+         :ok <- process_event_by_type(event),
+         :ok <- record_processed_event(event, customer_id) do
+      {:ok, event}
+    else
+      {:skip, reason} ->
+        Logger.info("Skipping stripe event", reason: inspect(reason))
+        {:ok, event}
 
       {:error, reason} ->
-        :ok =
-          Logger.error("Failed to sync account from Stripe",
-            customer_id: customer_id,
-            reason: inspect(reason)
-          )
+        Logger.error("Failed to process stripe event",
+          customer_id: customer_id,
+          reason: inspect(reason)
+        )
+
+        {:error, reason}
+
+      error ->
+        Logger.error("Unknown failure processing stripe event",
+          customer_id: customer_id,
+          reason: inspect(error)
+        )
 
-        :error
+        {:error, :unknown_failure}
     end
   end
 
-  # subscription is ended or deleted
-  def handle_event(%{
-        "object" => "event",
-        "data" => %{
-          "object" => %{
-            "customer" => customer_id
-          }
-        },
-        "type" => "customer.subscription.deleted"
-      }) do
-    update_account_by_stripe_customer_id(customer_id, %{
-      disabled_at: DateTime.utc_now(),
-      disabled_reason: "Stripe subscription deleted"
-    })
-    |> case do
-      {:ok, _account} ->
+  defp check_event_processing_eligibility(event, customer_id) do
+    %{
+      "id" => event_id,
+      "created" => event_created
+    } = event
+
+    cond do
+      ProcessedEvents.event_processed?(event_id) ->
+        {:skip, :already_processed}
+
+      true ->
+        check_chronological_order(customer_id, event_created, event_id)
+    end
```

Comment:
if / else seems clearer for only two clauses?

Reply:
Yeah, probably.  Credo agrees with you as well.